---
title: "SULF2 is a serological biomarker in MASLD"
author: "Christian Sommerauer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
  pdf_document:
    highlight: tango
subtitle: Differential gene expression analysis between two MASLD groups indentify
  11 biomarker candidate genes encoding secreted protein products.
---

# DESeq2 script to be performed with the public patient cohorts
#1 Govaere et al
#2 Kozumi et al

```{r setup, include=FALSE}
# set path
path <- strsplit(dirname(rstudioapi::getActiveDocumentContext()$path), '/')[[1]]
path <- path[-c((length(path)-0):length(path))]
path <- paste(path, collapse = '/')
knitr::opts_knit$set(root.dir = path)
knitr::opts_chunk$set(echo = T, message = F, warning = F, dev = 'png')
```

# Load the unfiltered, non-normalized (but formatted) data frames from the individual datasets
```{r}
#[[1]] = govaere, [[2]] = kozumi
getwd()
library(DESeq2)
library(tidyverse)

# Add data and metainfo into a list.
govaere = list(counts = readRDS("data/GSE135251/Processed/Govaere_processed.rds")[[1]] %>% dplyr::select(-Length) %>% tibble::rownames_to_column("GeneID"),
                              meta = read.delim("data/GSE135251/Processed/Govaere_meta.txt"))

# Add the AT_RISK classification and separate individuals into healthy and very early as well as early to moderate and severe.
govaere$meta <- govaere$meta %>% 
                    mutate(AT_RISK = case_when(
                    nas_score > 5 | Fibrosis_stage > 2 ~ "early_to_moderate_severe",                    
                    nas_score > 3 & Fibrosis_stage > 1 ~ "early_to_moderate_severe",                    
                    TRUE ~ "healthy_very_early")) %>%                    
                    mutate(Fibrosis_stage = as.factor(Fibrosis_stage)) %>%                    
                  mutate(nas_score = as.factor(nas_score))


table(govaere$meta$AT_RISK)

# Load the Kozumi data and add it to a list together with the govaere dataset.
df_list <- list(govaere = govaere,
                kozumi  = list(counts = readRDS("data/GSE167523/Processed/Kozumi_processed.rds")[[1]] %>% dplyr::select(-Length) %>% tibble::rownames_to_column("GeneID"),
                               meta = read.delim("data/GSE167523/Processed/Kozumi_meta.txt")))

# Inspect the list.
#df_list

# Group the df by the AT_RISK classifier for plotting.
group_by_at_risk <- df_list$govaere$meta %>% group_by(AT_RISK)

order_facet <- c("healthy_very_early", "early_to_moderate_severe")

# Plot patient distribution for fibrosis
ggplot(group_by_at_risk) +
  geom_histogram(aes(x=Fibrosis_stage), stat="count") +
  facet_wrap(~ factor(AT_RISK, levels=order_facet)) +
  theme_bw()

ggsave("results/240708_Fig1B_Govaere_Fibrosis_distribution.pdf", width=5, height=2)

# Plot patient distribution for NAS
ggplot(group_by_at_risk) +
  geom_histogram(aes(x=nas_score), stat="count") +
  facet_wrap(~ factor(AT_RISK, levels=order_facet)) +
  theme_bw()

ggsave("results/240708_Fig1B_Govaere_NAS_distribution.pdf", width=5, height=2)
```

```{r}
# Run the DESeq standard code in a loop for both datasets.
DESeq2_list <- list()
for (i in 1:length(df_list)) {
  df_list[[i]]$counts <- df_list[[i]]$counts %>%
    column_to_rownames("GeneID") %>%
    as.matrix()
  
DESeq2_list[[i]] <- DESeqDataSetFromMatrix(countData = df_list[[i]]$counts,
                                 colData = df_list[[i]]$meta,
                                 design = ~ 0 + AT_RISK)

DESeq2_list[[i]] <- estimateSizeFactors(DESeq2_list[[i]])
DESeq2_list[[i]] <- DESeq(DESeq2_list[[i]])

# filtering to minimum 100 counts in at least half of the individuals.
sample_number_threshold <- ncol(DESeq2_list[[i]])*0.5
DESeq2_list[[i]] <- DESeq2_list[[i]][rowSums(counts(DESeq2_list[[i]], normalized=TRUE) >= 100 ) >= sample_number_threshold, ]
}
names(DESeq2_list) <- names(df_list[])


# Specifiy the contrasts and add them into a list. [run only once, has to be run together with the previous 20 lines since objects are overwritten.]
DESeq2.DEGs <- list()

DESeq2.DEGs[[1]] <- list(
  no_vs_risk = results(DESeq2_list$govaere, contrast = c('AT_RISK', 'healthy_very_early', 'early_to_moderate_severe')))

DESeq2.DEGs[[2]] <- list(no_vs_risk = results(DESeq2_list$kozumi, contrast = c('AT_RISK', 'healthy_very_early', 'early_to_moderate_severe')))

names(DESeq2.DEGs) <- names(DESeq2_list)
```

# Add the SYMBOL annotations
```{r}
gene_ann <- read.delim("data/Annotations/230918_february_2023_GeneID_Symbol_Hsap.txt")
# add annotation to DEG lists

## For the Govaere et al dataset 
for (dataset in 1) { 
  for (contrast in 1) { 
DESeq2.DEGs[[dataset]][[contrast]] <- DESeq2.DEGs[[dataset]][[contrast]] %>% 
  as.data.frame() %>%
  tibble::rownames_to_column('GeneID') %>%
  inner_join(x=., y=gene_ann, by="GeneID")

 }
}

## For the Kozumi et al dataset 
DESeq2.DEGs$kozumi$no_vs_risk <- DESeq2.DEGs$kozumi$no_vs_risk %>% 
  as.data.frame() %>%
  tibble::rownames_to_column('GeneID') %>%
  inner_join(x=., y=gene_ann, by="GeneID")

names(DESeq2.DEGs)

# Add this to the complete gene list without filtering for the significant ones.
DESeq2.DEGs.unfiltered <- DESeq2.DEGs
```
# Filter the tables for foldchange and significance.
```{r}
# Filter both Govaere and Kozumi datasets by the indicated thresholds.
DESeq2.DEGs$govaere$no_vs_risk <- DESeq2.DEGs$govaere$no_vs_risk %>%
  dplyr::filter(padj < 1e-10  & log2FoldChange < -0.3785) # log2FC 0.3785 corresponds to 30 % upreg in disease.

DESeq2.DEGs$kozumi$no_vs_risk <- DESeq2.DEGs$kozumi$no_vs_risk %>%
  dplyr::filter(padj < 1e-10  & log2FoldChange < -0.3785) #  log2FC 0.3785 corresponds to 30 % upreg in disease.

# Add these filtered dataframes to the list.
DESeq2.DEGs.filtered <- DESeq2.DEGs

# These are the Foldchanges and significances for SULF2.
DESeq2.DEGs.filtered$govaere$no_vs_risk %>% filter(symbol=="SULF2")
DESeq2.DEGs.filtered$kozumi$no_vs_risk %>% filter(symbol=="SULF2")
```

# Check how many overlapping genes can be detectable in blood - and hence may serve as potential biomarker.
```{r}
# Load the table with proteins detectable in blood from the human protein atlas.
secreted.peps <- read.delim("data/HPA_MS_plasma/HPM_updated.txt") %>% rename("symbol" = "Gene")

# Make a separate dataframe for secreted proteins that are differentially expressed.
DESeq2.DEGs.filtered.secreted <- DESeq2.DEGs.filtered
# Overlap secreted genes with DEGs in the Kozumi et al dataset
DESeq2.DEGs.filtered.secreted$kozumi$no_vs_risk <- DESeq2.DEGs.filtered.secreted$kozumi$no_vs_risk %>%
    dplyr::filter(symbol %in% secreted.peps$symbol)
# Overlap secreted genes with DEGs in the Govaere et al dataset
 DESeq2.DEGs.filtered.secreted$govaere$no_vs_risk <- DESeq2.DEGs.filtered.secreted$govaere$no_vs_risk %>%
    dplyr::filter(symbol %in% secreted.peps$symbol)
# Intersect the secreted DEGs genes in the two datasets.
full_intersection_Datasets.secreted <- intersect(
  DESeq2.DEGs.filtered.secreted$govaere$no_vs_risk$symbol,
                                          DESeq2.DEGs.filtered.secreted$kozumi$no_vs_risk$symbol)


# Figure 1C
## Plot a Venn diagram with ggvenn
library(ggvenn)
library(ggplot2)

venn.list <- list(
Govaere.secr  = DESeq2.DEGs.filtered.secreted$govaere$no_vs_risk$symbol,
Kozumi.secr = DESeq2.DEGs.filtered.secreted$kozumi$no_vs_risk$symbol)

ggvenn(data=venn.list, c("Govaere.secr", "Kozumi.secr"), auto_scale = T, fill_color = c("black", "black"), show_percentage = F, text_size=10, fill_alpha = 0.4)
ggsave(file="results/ggvenn_koz_gov_1e_neg10_03785_secr.pdf")


# Supplementary Figure 1A
## Investigate the functions of the 11 shared genes using gene ontology.
library(clusterProfiler)
library(org.Hs.eg.db)

GO.result <- enrichGO(gene = full_intersection_Datasets.secreted,
  keyType = "SYMBOL",
  OrgDb = "org.Hs.eg.db",
  ont = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.05,
  qvalueCutoff = 0.05,
  minGSSize =3,
  readable = F)

head(GO.result)
#View(GO.result@result)

# Order the plots by p.adjust, then plot the 10 most significant terms.
plot_me_ordered <- GO.result[order(GO.result$p.adjust), ]
plot_me_ordered <- plot_me_ordered[1:10, ]
plot_me_ordered <- plot_me_ordered[order(plot_me_ordered$Count), ]
 name_order <- plot_me_ordered %>%
    dplyr::pull("Description")
  
ggplot(plot_me_ordered, aes(x=factor(Description, levels=name_order), fill=-log10(p.adjust), y=factor(""))) +
    geom_point(shape=21, aes(size=Count, fill=-log10(p.adjust))) +
    coord_flip() +
    scale_fill_gradient(low = "#808b96", high = "black", 
    limits = c(2, 6), breaks = c(2, 4, 6))+
    theme_classic() +
    theme(text=element_text(size = 18)) +
    ggtitle("") + 
    xlab("")

ggsave("results/GO_shared11.pdf", width = 18, height =10, units="cm")


# Supplementary Figure 1B
## Extract the 11 overlapping genes and plot their plasma concentration provided the table from the human protein atlas.
secrected_overlap <- intersect(venn.list$Kozumi.secr, venn.list$Govaere.secr)
secrected_overlap

secreted.peps_conc <- secreted.peps %>% filter(symbol %in% secrected_overlap)
# Multiply GSN by 1000 to get ug/L
secreted.peps_conc$Concentration[1] <- secreted.peps_conc$Concentration[1] * 1000
secreted.peps_conc$Unit[1] <- "ug/L"

ggplot(secreted.peps_conc, aes(x=symbol, y=log10(Concentration))) +
  geom_col(col="black", fill="grey") +
  geom_text(size = 6, aes(label=Concentration), vjust=1.5) +
  theme(axis.text=element_text(size=30)) +
  theme_minimal()
ggsave(file="results/biomarker_candidates_plasmaMS.pdf", width=25, height=10, units="cm")

# Save secreted proteins with plasma concentrations as a txt and all secreted DEGs in the lists as an R object.
write.table(secreted.peps_conc, "results/secreted_overlapping_1e_neg10_03785_koz_gov.txt", sep="\t", quote=F, row.names = F)
saveRDS(DESeq2.DEGs.filtered.secreted, "results/DESeq2_Govaere_Kozumi_1e_neg10_03785_secreted.rds")
```


```{r}
# Figures 1D-E
## Plot a volcano plot of all the secreted factors in both datasets

library(ggrepel)
## Create a dataframe with all expressed genes but not filtered for secreted protein-encoding genes; for the Govaere et al dataset.
DESeq2.DEGs.unfiltered_govaere <- DESeq2.DEGs.unfiltered$govaere$no_vs_risk %>% dplyr::filter(symbol %in% secreted.peps$symbol)
DESeq2.DEGs.govaere.label <- DESeq2.DEGs.filtered.secreted$govaere$no_vs_risk %>% filter(symbol %in% full_intersection_Datasets.secreted)

# Plot the volcano plot for Govaere et al dataset, label the overlapping 11 genes. Only genes that encode secreted protein products are plotted.
ggplot() +
  geom_point(data=DESeq2.DEGs.unfiltered_govaere, aes(x=-log2FoldChange, y=-log10(padj)), color="grey") +
  geom_point(data=DESeq2.DEGs.filtered.secreted$govaere$no_vs_risk,  aes(x=-log2FoldChange, y=-log10(padj)), color="red") +
  theme(axis.text=element_text(size=20)) +
  theme_minimal() +
  geom_text_repel(data=DESeq2.DEGs.govaere.label, aes(x=-log2FoldChange, y=-log10(padj)), label=full_intersection_Datasets.secreted, min.segment.length = unit(0, 'lines')) + 
  scale_x_continuous(limits=c(-2.4, 2.4), breaks=c(-2,-1,0,1,2))
ggsave(file="results/volcano_govaere_label.pdf", width=9, height=8, units="cm")

## Create a dataframe with all expressed genes but not filtered for secreted protein-encoding genes; for the Kozumi et al dataset.
DESeq2.DEGs.unfiltered_kozumi <- DESeq2.DEGs.unfiltered$kozumi$no_vs_risk %>%dplyr::filter(symbol %in% secreted.peps$symbol)
DESeq2.DEGs.kozumi.label <- DESeq2.DEGs.filtered.secreted$kozumi$no_vs_risk %>% filter(symbol %in% full_intersection_Datasets.secreted)

# Plot the volcano plot for Govaere et al dataset, label the overlapping 11 genes. Only genes that encode secreted protein products are plotted.
ggplot() +
  geom_point(data=DESeq2.DEGs.unfiltered_kozumi, aes(x=-log2FoldChange, y=-log10(padj)), color="grey") +
  geom_point(data=DESeq2.DEGs.filtered.secreted$kozumi$no_vs_risk,  aes(x=-log2FoldChange, y=-log10(padj)), color="red") +
  theme(axis.text=element_text(size=20)) +
  theme_minimal() +
  geom_text_repel(data=DESeq2.DEGs.kozumi.label, aes(x=-log2FoldChange, y=-log10(padj)), label=DESeq2.DEGs.kozumi.label$symbol, min.segment.length = unit(0, 'lines')) +
  scale_x_continuous(limits=c(-2.4, 2.4), breaks=c(-2,-1,0,1,2))
ggsave(file="results/volcano_kozumi_label.pdf", width=9, height=8, units="cm")
  
```
```{r}
sessionInfo()
```

